contracts/pools/yv2/StableSwapYv2.vy

```diff
@@ -0,0 +1,1009 @@
+    xp: uint256[N_COINS] = self._xp(rates)
+    precisions: uint256[N_COINS] = PRECISION_MUL
+
+    y: uint256 = xp[j] - (dy * FEE_DENOMINATOR / (FEE_DENOMINATOR - self.fee)) * precisions[j]
+    x: uint256 = self.get_y(j, i, y, xp)
+    dx: uint256 = (x - xp[i]) / precisions[i]
+    return dx
+
+
+@internal
+def _exchange(i: int128, j: int128, dx: uint256) -> uint256:
+    assert not self.is_killed
+    # dx and dy are in cy tokens
+
+    rates: uint256[N_COINS] = self._current_rates()
+    old_balances: uint256[N_COINS] = self.balances
+
+    xp: uint256[N_COINS] = empty(uint256[N_COINS])
+    for k in range(N_COINS):
+        xp[k] = rates[k] * old_balances[k] / PRECISION
+
+    x: uint256 = xp[i] + dx * rates[i] / PRECISION
+    y: uint256 = self.get_y(i, j, x, xp)
+    dy: uint256 = xp[j] - y - 1  # -1 just in case there were some rounding errors
+    dy_fee: uint256 = dy * self.fee / FEE_DENOMINATOR
+
+    dy = (dy - dy_fee) * PRECISION / rates[j]
+    dy_admin_fee: uint256 = dy_fee * self.admin_fee / FEE_DENOMINATOR
+    dy_admin_fee = dy_admin_fee * PRECISION / rates[j]
+
+    self.balances[i] = old_balances[i] + dx
+    self.balances[j] = old_balances[j] - dy - dy_admin_fee
+
+    return dy
+
+
+@external
+@nonreentrant('lock')
+def exchange(i: int128, j: int128, dx: uint256, min_dy: uint256) -> uint256:
+    """
+    @notice Perform an exchange between two coins
+    @dev Index values can be found via the `coins` public getter method
+    @param i Index value for the coin to send
+    @param j Index valie of the coin to recieve
+    @param dx Amount of `i` being exchanged
+    @param min_dy Minimum amount of `j` to receive
+    @return Actual amount of `j` received
+    """
+    dy: uint256 = self._exchange(i, j, dx)
+    assert dy >= min_dy, "Too few coins in result"
+
+    ERC20(self.coins[i]).transferFrom(msg.sender, self, dx)
+    ERC20(self.coins[j]).transfer(msg.sender, dy)
+
+    log TokenExchange(msg.sender, i, dx, j, dy)
+
+    return dy
+
+
+@external
+@nonreentrant('lock')
+def exchange_underlying(i: int128, j: int128, dx: uint256, min_dy: uint256) -> uint256:
+    """
+    @notice Perform an exchange between two underlying coins
+    @dev Index values can be found via the `underlying_coins` public getter method
+    @param i Index value for the underlying coin to send
+    @param j Index valie of the underlying coin to recieve
+    @param dx Amount of `i` being exchanged
+    @param min_dy Minimum amount of `j` to receive
+    @return Actual amount of `j` received
+    """
+
+    ERC20(self.underlying_coins[i]).transferFrom(msg.sender, self, dx)
+
+    coin: address = self.coins[i]
+
+    dx_: uint256 = ERC20(coin).balanceOf(self)
+    assert cyToken(coin).mint(dx) == 0
+    dx_ = ERC20(coin).balanceOf(self) - dx_
+    dy_: uint256 = self._exchange(i, j, dx_)
+
+    assert cyToken(self.coins[j]).redeem(dy_) == 0
+
+    underlying: address = self.underlying_coins[j]
+
+    dy: uint256 = ERC20(underlying).balanceOf(self)
```

Comment:
Just a very bizare idea (though it doesn't change too much). If we add `- 1` here, storage becomes cheaper, and we save 15k gas